accounts/abi/abi.libevm.go

```diff
@@ -0,0 +1,113 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser
+// General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see
+// <http://www.gnu.org/licenses/>.
+
+package abi
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/libevm/common"
+)
+
+// PackEvent packs the given `args` to conform with the ABI for the specified
+// event. Arguments MUST match the order specified in the event ABI. Indexed
+// arguments are returned as topics, as described in the [Solidity docs], while
+// the rest are packed into `data`.
+//
+// Struct, slice, and array arguments are not supported except for `[]byte`.
+//
+// [Solidity docs]: https://docs.soliditylang.org/en/latest/abi-spec.html#encoding-of-indexed-event-parameters
+func (abi ABI) PackEvent(name string, args ...any) (topics []common.Hash, data []byte, _ error) {
+	event, ok := abi.Events[name]
+	if !ok {
+		return nil, nil, fmt.Errorf("event %q not found", name)
+	}
+	if got, want := len(args), len(event.Inputs); got != want {
+		return nil, nil, fmt.Errorf("event %q received %d inputs; expecting %d", name, got, want)
+	}
+
+	var (
+		indexed  []any
+		packed   []any
+		packArgs Arguments
+	)
+	for i, arg := range args {
+		if inp := event.Inputs[i]; inp.Indexed {
+			indexed = append(indexed, arg)
+		} else {
+			packed = append(packed, arg)
+			packArgs = append(packArgs, inp)
+		}
+	}
+
+	topics, err := makeTopics1D(indexed)
+	if err != nil {
+		return nil, nil, err
+	}
+	if !event.Anonymous {
+		topics = append([]common.Hash{event.ID}, topics...)
+	}
+
+	data, err = packArgs.Pack(packed...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return topics, data, nil
+}
+
+func makeTopics1D(a []any) ([]common.Hash, error) {
+	t, err := MakeTopics(a)
+	if err != nil {
+		return nil, err
+	}
+	return t[0], nil
+}
+
+// PackOutput packs the given `args` to conform with the ABI for the specified
+// method's output.
+func (abi ABI) PackOutput(method string, args ...any) ([]byte, error) {
+	m, ok := abi.Methods[method]
+	if !ok {
+		return nil, fmt.Errorf("method %q not found", method)
+	}
+	return m.Outputs.Pack(args...)
+}
+
+// UnpackInputIntoInterface is equivalent to [ABI.UnpackIntoInterface] except
+// that it treats `data` as input when processing a method; all of the same
+// caveats apply.
+func (abi ABI) UnpackInputIntoInterface(v any, methodOrEventName string, data []byte) error {
+	in, err := abi.methodOrEventInputs(methodOrEventName)
+	if err != nil {
+		return err
+	}
+	unpacked, err := in.Unpack(data)
+	if err != nil {
+		return err
+	}
+	return in.Copy(v, unpacked)
+}
+
+func (abi ABI) methodOrEventInputs(name string) (Arguments, error) {
+	if m, ok := abi.Methods[name]; ok {
+		return m.Inputs, nil
+	}
+	if ev, ok := abi.Events[name]; ok {
+		return ev.Inputs, nil
```

Comment:
```suggestion
		return return ev.Inputs.NonIndexed(), nil
```

@iansuvak how do you use `UnpackInputIntoInterface()` for events? Packing and unpacking should generally allow for round-tripping but that only makes sense for non-indexed event arguments. Does this suggested change make sense?

Reply:
We do not use it directly -- your change here makes sense to me

Reply:
That being said it only makes sense to me in a vacuum. I don't know how anyone else might be using it but your comment about indexed vs non-indexed fields checks out and I think we should go with it.

Reply:
I'll just document it clearly.

Reply:
Turns out it actually does this under the hood anyway, so I haven't changed the implementation but I have added a test to demonstrate it.